src/placement/tests/pl_bench.c

```diff
@@ -0,0 +1,693 @@
+		type_idx = num_map_types - 1;
+		percent_moved[type_idx * (domains_to_add + 1) + added] =
+			(double)added * 1 * nodes_per_domain /
+			(1 * nodes_per_domain * num_domains +
+			 added * 1 * nodes_per_domain);
+	}
+
+	/* Print out the data */
+	for (type_idx = 0; type_idx < num_map_types; type_idx++) {
+		D_PRINT("Addition Data: Type %d\n", type_idx);
+		for (added = 0; added <= domains_to_add; added++) {
+			D_PRINT("%f\n",
+				percent_moved[type_idx * (domains_to_add + 1)
+						       + added]);
+		}
+	}
+	D_PRINT("\n");
+
+	benchmark_graph((double *)percent_moved, map_keys, num_map_types,
+			domains_to_add + 1, "Number of added racks",
+			"% Data Moved", 1.0,
+			"Data movement %% when adding racks", "/tmp/gnufifo",
+			use_x11);
+}
+
+
+int
+main(int argc, char **argv)
+{
+	uint32_t                 num_domains = DEFAULT_NUM_DOMAINS;
+	uint32_t                 nodes_per_domain = DEFAULT_NODES_PER_DOMAIN;
+	uint32_t                 vos_per_target = DEFAULT_VOS_PER_TARGET;
+
+	int                      rc;
+	int                      i;
+
+	test_op_t operation = NULL;
+
+	test_op_t op_fn[] = {
+		benchmark_placement,
+		benchmark_add_data_movement,
+	};
+	const char *const op_names[] = {
+		"benchmark-placement",
+		"benchmark-add",
+	};
+	D_ASSERT(ARRAY_SIZE(op_fn) == ARRAY_SIZE(op_names));
+
+	while (1) {
+		static struct option long_options[] = {
+			{"operation", required_argument, 0, 'o'},
+			{"num-domains", required_argument, 0, 'd'},
+			{"nodes-per-domain", required_argument, 0, 'n'},
+			{"vos-per-target", required_argument, 0, 'v'},
+			{"gdb-wait", no_argument, 0, 'g'},
+			{0, 0, 0, 0}
+		};
+		int c;
+		int ret;
+
+		c = getopt_long(argc, argv, "o:d:n:v:g", long_options, NULL);
+		if (c == -1)
+			break;
+
+		switch (c) {
+		case 'd':
+			ret = sscanf(optarg, "%u", &num_domains);
+			if (ret != 1) {
+				num_domains = DEFAULT_NUM_DOMAINS;
+				D_PRINT("Warning: Invalid num-domains\n"
+					"  Using default value %u instead\n",
+					num_domains);
+			}
+			break;
+		case 'n':
+			ret = sscanf(optarg, "%u", &nodes_per_domain);
+			if (ret != 1) {
+				nodes_per_domain = DEFAULT_NODES_PER_DOMAIN;
+				D_PRINT("Warning: Invalid nodes-per-domain\n"
+					"  Using default value %u instead\n",
+					nodes_per_domain);
+			}
+			break;
+		case 'v':
+			ret = sscanf(optarg, "%u", &vos_per_target);
+			if (ret != 1) {
+				vos_per_target = DEFAULT_VOS_PER_TARGET;
+				D_PRINT("Warning: Invalid vos-per-target\n"
+					"  Using default value %u instead\n",
+					vos_per_target);
+			}
+			break;
+		case 'o':
+			for (i = 0; i < ARRAY_SIZE(op_fn); i++) {
+				if (strncmp(optarg, op_names[i],
+					    strlen(op_names[i])) == 0) {
+					operation = op_fn[i];
+					break;
+				}
+			}
+			if (i == ARRAY_SIZE(op_fn)) {
+				D_PRINT("ERROR: Unknown operation '%s'\n",
+					optarg);
+				print_usage(argv[0], op_names,
+					    ARRAY_SIZE(op_names));
+				return -1;
+			}
+			break;
+		case 'g':
+			{
+				volatile int gdb = 0;
```

Comment:
(style)  Use of volatile is usually wrong: see Documentation/volatile-considered-harmful.txt

Reply:
Volatile usage is correct in this case - this code is also not executed under normal circumstances. If it is not marked volatile, the compiler will optimize it out - since it is never set to non-zero here (it is intended to be set with GDB)